theano/tensor/basic.py

```diff
@@ -6306,7 +6306,22 @@ def __str__(self):
 )
 
 
-def dot(a, b):
+def dot(l, r):
+    """Return a symbolic dot product.
+    This is designed to work with both sparse and dense tensors types.
+    """
+    try:
+        res = l.__dot__(r)
```

Comment:
What if `l.__dot__` _returns_ `NotImplemented`; wouldn't we want to try `r.__rdot__`?

Reply:
Hmm, that's not what the old implementation did - if `l.__dot__(r)` returned `NotImplemented`, then it would just run a `NotImplementedError`. Is that behavior we'd want? I can't imagine a situation where one returned `NotImplemented` while the other didn't.

Reply:
The old `theano.dot` would check for a return value of `NotImplemented` and raise its own `NotImplementedError` so that the `r.__rdot__` call in the `except` would be tried (see [this statement](https://github.com/pymc-devs/Theano-PyMC/blob/master/theano/__init__.py#L171)).

Reply:
Ah whoops sorry, don't know how I missed that. The next commit should fix this.